File: IOExplorer/Summaries/SummaryNode.m
//
//  SummaryNode.m
//  IOExplorer
//
//  Created by Jinwoo Kim on 6/5/23.
//

#import "SummaryNode.h"

@interface SummaryNode ()
@property (copy) NSString *title;
@property (copy) NSString *systemImageName;
@property (copy) NSArray<SummaryNode *> *children;
@end

@implementation SummaryNode

- (instancetype)initWithTitle:(NSString *)title systemImageName:(NSString *)systemImageName children:(NSArray<SummaryNode *> *)children {
    if (self = [self init]) {
        self.title = title;
        self.systemImageName = systemImageName;
        self.children = children;
    }
    
    return self;
}

- (void)dealloc {
    [_title release];
    [_systemImageName release];
    [_children release];
    [super dealloc];
}

- (id)copyWithZone:(struct _NSZone *)zone {
    SummaryNode *copy = [self.class new];
    
    if (copy) {
        copy.title = self.title;
        copy.systemImageName = self.systemImageName;
        copy.children = self.children;
    }
    
    return copy;
}

- (BOOL)isLeaf {
    return !self.children.count;
}

- (NSNumber *)count {
    return @(self.children.count);
}

@end

File: IOExplorer/Summaries/SummaryNode.cpp
//
//  SummaryNode.cpp
//  IOExplorer
//
//  Created by Jinwoo Kim on 6/5/23.
//

#import "SummaryNode.hpp"
#import <IOKit/usb/IOUSBLib.h>

SummaryNode::SummaryNode(io_object_t ioObject, std::string title, std::vector<std::shared_ptr<SummaryNode>> children) : _ioObject(ioObject), _title(title), _children(children) {
    IOObjectRetain(ioObject);
}

SummaryNode::~SummaryNode() {
    IOObjectRelease(_ioObject);
}

io_object_t SummaryNode::ioObject() {
    return _ioObject;
}

std::string SummaryNode::title() {
    return _title;
}

std::string SummaryNode::systemImageName() {
    if (_ioObject == IO_OBJECT_NULL) {
        return "questionmark";
    }
    
    CFStringRef classNameRef = IOObjectCopyClass(_ioObject);
    
    if (CFStringCompare(classNameRef, CFSTR(kIOUSBHostDeviceClassName), 0) == kCFCompareEqualTo) {
        return "heart";
    } else {
        return "questionmark";
    }
}

std::vector<std::shared_ptr<SummaryNode>> SummaryNode::children() {
    return _children;
}

bool SummaryNode::isLeaf() {
    return _children.empty();
}

SummaryNode::SummaryNode(const SummaryNode& other) : _ioObject(other._ioObject), _title(other._title), _children(other._children) {
    IOObjectRetain(_ioObject);
}

SummaryNode& SummaryNode::operator=(const SummaryNode &other) {
    if (this != &other) {
        _ioObject = other._ioObject;
        _title = other._title;
        _children = other._children;
        IOObjectRetain(_ioObject);
    }
    
    return *this;
}
